reactive/kotlinx-coroutines-reactor/src/Flux.kt

```diff
@@ -32,9 +42,238 @@ import kotlin.coroutines.*
  * **Note: This is an experimental api.** Behaviour of publishers that work as children in a parent scope with respect
  *        to cancellation and error handling may change in the future.
  */
+
 @ExperimentalCoroutinesApi
 fun <T> CoroutineScope.flux(
     context: CoroutineContext = EmptyCoroutineContext,
     @BuilderInference block: suspend ProducerScope<T>.() -> Unit
-): Flux<T> =
-    Flux.from(publish(newCoroutineContext(context), block = block))
+): Flux<T> = Flux.from(reactorPublish(newCoroutineContext(context), block = block))
+
+@ExperimentalCoroutinesApi
+public fun <T> CoroutineScope.reactorPublish(
+    context: CoroutineContext = EmptyCoroutineContext,
+    @BuilderInference block: suspend ProducerScope<T>.() -> Unit
+): Publisher<T> = Publisher { subscriber ->
+    // specification requires NPE on null subscriber
+    if (subscriber == null) throw NullPointerException("Subscriber cannot be null")
+    val currentContext = if (subscriber is CoreSubscriber) subscriber.currentContext() else Context.empty()
+    val reactorContext = (coroutineContext[ReactorContext]?.context?.putAll(currentContext) ?: currentContext).asCoroutineContext()
+    val newContext = newCoroutineContext(context + reactorContext)
+    val coroutine = PublisherCoroutine(newContext, subscriber)
+    subscriber.onSubscribe(coroutine) // do it first (before starting coroutine), to avoid unnecessary suspensions
+    coroutine.start(CoroutineStart.DEFAULT, coroutine, block)
+}
+
+private const val CLOSED = -1L    // closed, but have not signalled onCompleted/onError yet
+private const val SIGNALLED = -2L  // already signalled subscriber onCompleted/onError
+
+@Suppress("CONFLICTING_JVM_DECLARATIONS", "RETURN_TYPE_MISMATCH_ON_INHERITANCE")
+private class PublisherCoroutine<in T>(
```

Comment:
Copy-pasting the whole integration is definitely not an option: it is hard to maintain, hard to test and easy to mess up.
For example, we can make it public and mark with `InternalCoroutinesApi`